base/initdefs.jl

```diff
@@ -378,18 +378,26 @@ LIFO order, "last called" is equivalent to "first registered".)
 atexit(f::Function) = Base.@lock _atexit_hooks_lock (pushfirst!(atexit_hooks, f); nothing)
 
 function _atexit(exitcode::Cint)
-    while !isempty(atexit_hooks)
-        f = popfirst!(atexit_hooks)
-        try
-            if hasmethod(f, (Cint,))
-                f(exitcode)
-            else
-                f()
+    # Don't hold the lock around the iteration, just in case any other thread executing in
+    # parallel tries to register a new atexit hook while this is running. We don't want to
+    # block that thread from proceeding, and we can allow it to register its hook which we
+    # will immediately run here.
+    # NOTE: There is a logical race condition where a Task that registers a hook after we
+    # have finished will never have its hook run. We make no guarantees about this.
```

Comment:
@vtjnash: The other alternative would be to change some global atomic boolean to "exiting", so that anyone who attempts to register an atexit hook _after we've finished the last one_ will throw an error rather than silently succeeding, despite us ignoring their hook...

Part of me thinks that's a clearer and safer behavior. Presumably they can then catch that exception and clean up the resource right there.

Do you also prefer that?

Reply:
I'll push that up in another commit, and we can always drop it if we don't like it.

Reply:
Okay, done in [e31b144](https://github.com/JuliaLang/julia/pull/49868/commits/e31b1445fbe722c30478d4e7f43dac4e3f1678dc). Can you give me your preference?

Reply:
Yes, that seems a reasonable way to help alert people to the issue. Commonly `atexit` is only registered once, and then work is managed outside it (e.g. see tempfile cleanup)

Reply:
Ah, we could change our code to follow that pattern, i suppose. What is the advantage over registering one hook per object to clean up?